### PR TITLE
[Merged by Bors] - feat(CategoryTheory): Guitart exact squares

### PR DESCRIPTION
**Pull request by joelriou**
This PR defines the categorical notion of `2`-exact square in the sense of Guitart. It is introduced in order to prepare for future applications to derived functors.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/GuitartExact.lean [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+`w : TwoSquare T L R B` and `g : R.obj X₂ ⟶ B.obj X₃`, a category `J` is introduced
+and it is observed that it is equivalent to the two categories
+`w.JRightwards g` and `w.JDownwards g`. We shall show below
+that there is an equivalence `w.equivalenceJ g : w.JRightwards g ≌ w.JDownwards g`. -/
+
+/-- Given `w : TwoSquare T L R B` and a morphism `g : R.obj X₂ ⟶ B.obj X₃`, this is the
+category `StructuredArrow (CostructuredArrow.mk g) (w.costructuredArrowRightwards X₃)`,
+see the constructor `JRightwards.mk` for the data that is involved.-/
+abbrev JRightwards :=
+  StructuredArrow (CostructuredArrow.mk g) (w.costructuredArrowRightwards X₃)
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/GuitartExact.lean:97]: Is there a reason to include `J` in the names, beyond the fact that `J` was used in the paper you cite?

File: Mathlib/CategoryTheory/GuitartExact.lean [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+lemma isConnected_rightwards_iff_downwards :
+    IsConnected (w.JRightwards g) ↔ IsConnected (w.JDownwards g) :=
+  isConnected_iff_of_equivalence (w.equivalenceJ g)
+
+end
+
+/-- Condition on `w : TwoSquare T L R B` expressing that it is a Guitart exact square.
+See lemmas `guitartExact_iff_isConnected_rightwards`, `guitartExact_iff_isConnected_downwards`,
+`guitartExact_iff_final`, `guitartExact_iff_initial` for various characterizations. -/
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/GuitartExact.lean:168]: Please spell out at least one of these conditions/characterizations. Also, maybe provide a bit of motivation for this notion (as I think it isn't that well-known), or maybe provide a pointer to more info.

File: docs/references.bib [hunk 3]
```diff
@@ -1368,6 +1368,21 @@ @Unpublished{     grinberg_clifford_2016
   year          = {2016}
 }
 
+@Article{         Guitard1980,
```
Comment (jcommelin) [inline docs/references.bib:1371]: I guess this is a typo?
```suggestion
@Article{         Guitart1980,
```

File: Mathlib/CategoryTheory/GuitartExact.lean [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+## TODO
+
+* Define the notion of derivability structure from
+[the paper by Kahn and Maltsiniotis][KahnMaltsiniotis2008] using Guitart exact squares
+and construct (pointwise) derived functors using this notion
+
+## References
+* https://ncatlab.org/nlab/show/exact+square
+* [René Guitart, *Relations et carrés exacts*][Guitard1980]
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/GuitartExact.lean:38]:
```suggestion
* [René Guitart, *Relations et carrés exacts*][Guitart1980]
```

File: docs/references.bib (hunk 3 above)
Comment (joelriou) [reply docs/references.bib:1371]: Well spotted!

File: Mathlib/CategoryTheory/GuitartExact.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/GuitartExact.lean]: I have renamed this one as `StructuredArrowRightwards` and the other one as `CostructuredArrowDownwards`.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**